Create Ignition plugins for TeleportIngestor and TeleportDispenser and refactor with existing Gazebo plugins

mrushyendra opened the pull request:
Creates new `TeleportIngestorPlugin` and `TeleportDispenserPlugin` for Ignition. Tested with the Office Demo. 

Refactors both the Ignition and Gazebo plugins in order to move common state variables, callbacks and ROS message publishers/subscribers into common `TeleportIngestorCommon` and `TeleportDispenserCommon` classes. For the existing
Gazebo plugins, this involves setting new flags when Ingestor/Dispenser requests are made, and moving the actual ingesting/dispensing to the `on_update()` calls instead.

luca-della-vedova commented:
Can you try to merge / rebase to the latest master? Would like to check if github actions run for the linter

mrushyendra commented:
Sure, will do in the next few days. This might take quite a while because of recent changes to master.

luca-della-vedova commented:
I merged the latest master to reduce the size of the diff, can you rebase on top of the latest `feature/add_ignition_support`?

We should probably just have opened a PR for that branch into master anyway, will do after this gets in.

luca-della-vedova commented on rmf_demo_plugins/rmf_plugins_common/src/ingestor_common.cpp:
```diff
@@ -0,0 +1,92 @@
+
+namespace rmf_ingestor_common {
+
+TeleportIngestorCommon::TeleportIngestorCommon()
+{
+}
```
It seems there is quite a few cases of constructors / destructors being empty. I would suggest either changing them to
```
public:
  ObjectName() = default;
```
if you want to be explicit about them being default constructors or omit them altogether.

luca-della-vedova commented on rmf_demo_plugins/rmf_plugins_common/src/ingestor_common.cpp:
```diff
@@ -0,0 +1,92 @@
+  _fleet_state_sub = ros_node->create_subscription<FleetState>(
+    "/fleet_states",
+    rclcpp::SystemDefaultsQoS(),
+    [&](FleetState::UniquePtr msg)
+    {
+      fleet_state_cb(std::move(msg));
+    });
```
Since all the lambda does is call the callback can we pass the callback straight away instead?
Same in `dispenser_common.cpp`

luca-della-vedova reviewed (changes requested):
I generally see a fair bit of code duplication all around, for example the common ingestor and dispenser are very similar, as well as the plugins.
For example both the ingestor and dispenser have the same `find_nearest_non_static_model` function.

I would recommend refactoring to reduce code duplication as much as possible, trying to also push as much similar code between ignition and gazebo to the common class as possible. Ideally the gazebo / ignition plugins should only be thin wrappers that do the functions specific to the simulators, similar to what the [door](https://github.com/osrf/traffic_editor/blob/master/building_sim_plugins/building_gazebo_plugins/src/door.cpp) or the [slotcar](https://github.com/osrf/traffic_editor/blob/master/building_sim_plugins/building_gazebo_plugins/src/slotcar.cpp) plugins do.

The code will maybe not be super clean, with the common class becoming quite complex but it helps in maintaining since every change for the behavior of the plugins has to be done in only one file (the common class) and we don't have to worry about the plugins code.

mrushyendra commented on rmf_demo_plugins/rmf_gazebo_plugins/src/TeleportDispenser.cpp:
```diff
@@ -0,0 +1,273 @@
+
+  _dispenser_common->sim_time = _world->SimTime().Double();
+
+  // `_dispense` is set to true if the dispenser plugin node has received a valid DispenserRequest
+  if (_dispenser_common->dispense)
+  {
+    _dispenser_common->send_dispenser_response(DispenserResult::ACKNOWLEDGED);
+
+    if (_dispenser_common->dispenser_filled)
+    {
+      RCLCPP_INFO(_dispenser_common->ros_node->get_logger(),
+        "Dispensing item");
+      dispense_on_nearest_robot(_dispenser_common->latest.transporter_type);
+
+      _dispenser_common->send_dispenser_response(DispenserResult::SUCCESS);
+    }
+    else
+    {
+      RCLCPP_WARN(_dispenser_common->ros_node->get_logger(),
+        "No item to dispense: [%s]", _dispenser_common->latest.request_guid);
+      _dispenser_common->send_dispenser_response(DispenserResult::FAILED);
+    }
+    _dispenser_common->dispense = false;
+  }
```
I considered this earlier, but it involves making a call back to the `dispense_on_nearest_robot` function, and I wasn't sure if calling functions from the individual classes inside the common class would make things too complex. Also, since both `dispense_on_nearest_robot` functions take different arguments, I _think_ this may be possible using `std::bind` and `std::function`?

Either way as you said the savings from future maintenance and updates probably outweighs those concerns. Will look at moving it!

mrushyendra commented on rmf_demo_plugins/rmf_ignition_plugins/package.xml:
```diff
@@ -0,0 +1,32 @@
+<package format="3">
+  <name>rmf_ignition_plugins</name>
+  <version>1.0.0</version>
+  <description>
+    ROS 2 slotcar plugin and payload handlers for Gazebo simulated robots
+  </description>
+  <author>Morgan Quigley</author>
+  <author>Michael Grey</author>
+  <author>Roselle Carmen</author>
+  <author>Aaron Chong</author>
+  <author>Brandon Ong</author>
+  <author>Rushyendra Maganty</author>
+  <maintainer email="morgan@openrobotics.org">Morgan Quigley</maintainer>
+  <license>Apache 2.0</license>
```
Sure, will do. I actually got the information from the current rmf_gazebo_plugins package.xml file on master, somehow the description slipped my notice.

mrushyendra commented on rmf_demo_plugins/rmf_ignition_plugins/CMakeLists.txt:
```diff
@@ -0,0 +1,117 @@
+
+add_library(teleport_ingestor SHARED src/TeleportIngestor.cpp)
+
+target_link_libraries(teleport_ingestor
+  PUBLIC
+    ${rclcpp_LIBRARIES}
+    ignition-gazebo${IGN_GAZEBO_VER}::ignition-gazebo${IGN_GAZEBO_VER}
+    ignition-plugin${IGN_PLUGIN_VER}::ignition-plugin${IGN_PLUGIN_VER}
+    ${rmf_plugins_common_LIBRARIES}
+    ${rmf_fleet_msgs_LIBRARIES}
+    ${rmf_ingestor_msgs_LIBRARIES}
+)
```
Sure, changed all the CMakeLists files to use `ament_target_dependencies` instead.

mrushyendra commented on rmf_demo_plugins/rmf_ignition_plugins/src/TeleportDispenser.cpp:
```diff
@@ -0,0 +1,368 @@
+    {
+      RCLCPP_INFO(_dispenser_common->ros_node->get_logger(),
+        "Dispensing item");
+      dispense_on_nearest_robot(ecm,
+        _dispenser_common->latest.transporter_type);
+
+      _dispenser_common->send_dispenser_response(DispenserResult::SUCCESS);
```
The current tasks don't seem to behave any differently regardless of whether we return `DispenserResult::SUCCESS` or `DispenserResult::FAILED`. I went with following how the current Gazebo plugin did things, but I could change `dispense_on_nearest_robot` to return a bool instead, and then return `SUCCESS` or `FAIL` accordingly.

mrushyendra commented on rmf_demo_plugins/rmf_plugins_common/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+find_package(ignition-math4 REQUIRED)
```
For the current limited use case it works fine, but guess I'll bump it up ignition-math6 instead.

mrushyendra commented:
I further refactored the code and addressed the review comments. The code is structured such that the main ingesting/dispensing control flow, message passing and state variables are all contained in `TeleportDispenserCommon` and `TeleportIngestorCommon`, while functions that deal with simulator-specific physical movements remain in the individual `rmf_ignition_plugins` and `rmf_gazebo_plugins` folder. A byproduct of this is that the code is slightly more verbose in parts - mostly when binding functions to pass as callbacks - though duplication is reduced.

I also considered moving some functions that are similar between both the Ignition plugins, or between both the Gazebo plugins to some common library, e.g. `ignition_utils.hpp`, but this may make the code even more complex because it would involve passing a lot of class member variables to the functions, and would also make defining callback functions to pass to `on_update()` harder. Would love to hear your opinion on this - I can change it if needed.

luca-della-vedova commented on rmf_demo_plugins/rmf_plugins_common/CMakeLists.txt:
```diff
@@ -0,0 +1,114 @@
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+find_package(gazebo REQUIRED)
+find_package(gazebo_ros REQUIRED)
+find_package(gazebo_dev REQUIRED)
```
Can we get rid of the gazebo dependency for the utils library?

luca-della-vedova reviewed (commented):
I like the approach with function objects being passed to the common class, it makes the logic in the dispenser classes very nice and readable! Just the small comment on cleaning the dependencies in the common class then I think this is good to go

mrushyendra commented:
Sure, done!

luca-della-vedova reviewed (approved):
Thanks for all the changes! Looks good, will merge
